src/libertem/udf/base.py

```diff
@@ -1,6 +1,275 @@
+        """
+        raise NotImplementedError()
+
+
+class UDFTileMixin:
+    def process_tile(self, tile):
+        """
+        Implement this method to process the data in a tiled manner.
+
+        Data available in this method:
+        - `self.params`    - the parameters of this UDF
+        - `self.task_data` - task data created by `get_task_data`
+        - `self.results`   - the result buffer instances
+
+        Parameters
+        ----------
+        tile : ndarray
+            A small number N of frames or signal elements from the dataset.
+            The shape is (N,) + `dataset.shape.sig`. In case of pixelated
+            STEM / scanning diffraction data this is 3D, for spectra 2D etc.
+        """
+        raise NotImplementedError()
+
+
+class UDFPartitionMixin:
+    def process_partition(self, partition):
+        """
+        Implement this method to process the data partitioned into large
+        (100s of MiB) partitions.
+
+        Data available in this method:
+        - `self.params`    - the parameters of this UDF
+        - `self.task_data` - task data created by `get_task_data`
+        - `self.results`   - the result buffer instances
+
+        Note
+        ----
+        Only use this method if you know what you are doing; especially if
+        you are running a processing pipeline with multiple steps, or multiple
+        processing pipelines at the same time, performance may be adversely
+        impacted.
+
+        Parameters
+        ----------
+        partition : ndarray
+            A large number N of frames or signal elements from the dataset.
+            The shape is (N,) + `dataset.shape.sig`. In case of pixelated
+            STEM / scanning diffraction data this is 3D, for spectra 2D etc.
+        """
+        raise NotImplementedError()
+
+
+class UDFBase:
+    def allocate_for_part(self, partition, roi):
+        for ns in [self.results]:
+            ns.allocate_for_part(partition, roi)
+
+    def allocate_for_full(self, dataset, roi):
+        for ns in [self.results]:
+            ns.allocate_for_full(dataset, roi)
+
+    def set_views_for_partition(self, partition):
+        for ns in [self.params, self.results]:
+            ns.set_view_for_partition(partition)
+
+    def set_views_for_tile(self, partition, tile):
+        for ns in [self.params, self.results]:
+            ns.set_view_for_tile(partition, tile)
+
+    def set_views_for_frame(self, partition, tile, frame_idx):
+        for ns in [self.params, self.results]:
+            ns.set_view_for_frame(partition, tile, frame_idx)
+
+    def clear_views(self):
+        for ns in [self.params, self.results]:
+            ns.clear_views()
+
+    def init_task_data(self, partition):
+        self.task_data = UDFNamespace(self.get_task_data(partition))
+
+    def init_result_buffers(self):
+        self.results = UDFNamespace(self.get_result_buffers())
+
+
+class UDF(UDFBase):
+    def __init__(self, **kwargs):
```

Comment:
Draft: This method is called again on the cluster to create copies. For that reason it should not have side effects beyond `self` and only depend on its parameters.

Discussion with @sk1p: Overloading this method with sth fancy may have unwanted side effects.

Documentation: Include a clear example to condition parameters.